File: go/fib.go
// https://leetcode.com/problems/fibonacci-number/

func fib(n int) int {
	cache := make(map[int]int)

	var helper func(int) int
	helper = func(n int) int {
		if n < 2 {
			return n
		}

		if val, ok := cache[n]; ok {
			return val
		}

		cache[n] = helper(n-1) + helper(n-2)
		return cache[n]
	}

	return helper(n)
}

File: python/src/fib.py
# https://leetcode.com/problems/fibonacci-number/


class Solution:
    def fib(self, n: int) -> int:
        res = n
        if n < 2:
            return res

        n2, n1 = 0, 1
        for _ in range(2, n + 1):
            res = n2 + n1
            n2, n1 = n1, res

        return res
